workflow/controller/steps.go

```diff
@@ -170,19 +170,21 @@ func (woc *wfOperationCtx) executeSteps(ctx context.Context, nodeName string, tm
 	if err != nil {
 		return node, err
 	}
-
 	if outputs != nil {
+		node, err := woc.wf.GetNodeByName(nodeName)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
per my root cause analysis, we can have a more targeted fix by just reverting this part (in both steps and DAGs), so that we retain the memoization fix below. then I can solve the root cause in a follow-up PR

Reply:
I've done this in #12203

#12130 should not have gone in to 3.4 branch at all (as memoization without outputs was a feat), so this revert should be done on 3.4, but #12203 should be in 3.5.